shapes/main.cpp

```diff
@@ -1,79 +1,88 @@
-}
+};
 
-bool perimeterBiggerThan20(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getPerimeter() > 20);
-    return false;
-}
+auto perimeterBiggerThan20 = [](shared_ptr<Shape> s) { return (s) ? (s->getPerimeter() > 20) : (false); };
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
-    return false;
-}
+auto areaLessThanX = [x{10}](shared_ptr<Shape> s) { return (s) ? (s->getArea() < x) : (false); };
 
-void printCollectionElements(const Collection& collection)
-{
-    for(Collection::const_iterator it = collection.begin(); it != collection.end(); ++it)
-        if(*it)
-            (*it)->print();
+void printCollection(const Collection& collection) {
+    for (const auto& it : collection)
+        if (it)
+            it->print();
 }
 
-void printAreas(const Collection& collection)
-{
-    for(vector<shared_ptr<Shape>>::const_iterator it = collection.begin(); it != collection.end(); ++it)
-        if(*it)
-            cout << (*it)->getArea() << std::endl;
+void printAreas(const Collection& collection) {
+    for (const auto& it : collection)
+        if (it)
+            cout << it->getArea() << std::endl;
 }
 
 void findFirstShapeMatchingPredicate(const Collection& collection,
-                                     bool (*predicate)(shared_ptr<Shape> s),
-                                     std::string info)
-{
-    Collection::const_iterator iter = std::find_if(collection.begin(), collection.end(), predicate);
-    if(*iter != nullptr)
-    {
+                                     std::function<bool(shared_ptr<Shape> s)> predicate,
+                                     std::string info) {
+    auto iter = std::find_if(collection.begin(), collection.end(), predicate);
+    if (*iter) {
         cout << "First shape matching predicate: " << info << endl;
         (*iter)->print();
-    }
-    else
-    {
+    } else {
         cout << "There is no shape matching predicate " << info << endl;
     }
 }
 
-int main()
-{
-    Collection shapes;
-    shapes.push_back(make_shared<Circle>(2.0));
-    shapes.push_back(make_shared<Circle>(3.0));
-    shapes.push_back(nullptr);
-    shapes.push_back(make_shared<Circle>(4.0));
-    shapes.push_back(make_shared<Rectangle>(10.0, 5.0));
-    shapes.push_back(make_shared<Square>(3.0));
-    shapes.push_back(make_shared<Circle>(4.0));
-    printCollectionElements(shapes);
+constexpr int fibo(int n) {
+    if (n <= 2) {
+        return 1;
+    } else {
+        return fibo(n - 1) + fibo(n - 2);
+    }
+}
+
+// Task 11 - Sfinae
+template <typename T, typename = typename std::enable_if<std::is_base_of<Shape, T>::value>>
+void insertOnlyShapes(Collection& collection, std::shared_ptr<T> shape) {
+    collection.emplace_back(shape);
+}
+
+// Task 18 - Variadic template
+template <class DerivedType, class... Arguments>
+std::shared_ptr<Shape> make_shape(Arguments&&... args) {
+    return make_shared<DerivedType>(decltype(args)(args)...);
```

Comment:
the best way is to use perfect forwarding with std::forward. Without it you always pass l-value references.
```suggestion
    return make_shared<DerivedType>(std::forward<Arguments>(args)...);
```

Reply:
Thank you for letting me know, It is still magic for me :)